### PR TITLE
Add `EventListener` functions for write stalls

### PR DESCRIPTION
**Pull request by ajkr**
Call `WriteStallBegin()` when entering write stall mode and
`WriteStallEnd()` when leaving write stall mode.

### CODE REVIEW

**Comment by petermattis** [comment]
This change is [Reviewable](https://reviewable.io/reviews/petermattis/pebble/209)

**Comment by ajkr** [comment]
^also should investigate why the first flush took 12 seconds.

**Comment by Ryanfsdf** [comment]
> ^also should investigate why the first flush took 12 seconds.

I wonder if this has to do with my memtable flushing mechanism (though the mechanism shouldn't throttle flush rate when at the memtable limit)

**Comment by ajkr** [comment]
> I'm not sure if the write stall is ended at this point. I think you need a loop:

Good point, what I was thinking.

> That's just my idea for a test. Feel free to suggest something better.

Not sure if better, but I think we could do this entirely with a custom `EventListener` by configuring its `FlushBegin()` callback to sleep until `WriteStallBegin()` is called. Then flush can be released to proceed and we verify `WriteStallEnd()` is eventually called. Will try it out.

**Comment by petermattis** [comment]
> Not sure if better, but I think we could do this entirely with a custom EventListener by configuring its FlushBegin() callback to sleep until WriteStallBegin() is called. Then flush can be released to proceed and we verify WriteStallEnd() is eventually called. Will try it out.

I think that should work. That is a better idea.

**Comment by ajkr** [comment]
> > Not sure if better, but I think we could do this entirely with a custom EventListener by configuring its FlushBegin() callback to sleep until WriteStallBegin() is called. Then flush can be released to proceed and we verify WriteStallEnd() is eventually called. Will try it out.
> 
> I think that should work. That is a better idea.

Hm, it's getting kind of complicated because the DB mutex is held while `FlushBegin()` runs. That'd prevent writes to the second memtable or initiating the second flush which prevents us from hitting stall mode. Delaying in the manifest write sounds better now.

File: db.go [hunk 1]
```diff
@@ -735,33 +735,31 @@ func (d *DB) manualCompact(manual *manualCompaction) error {
...
-	<-mem.flushed()
+	<-flushDone
 	return nil
 }
 
 // AsyncFlush asynchronously flushes the memtable to stable storage.
-//
-// TODO(peter): untested
-func (d *DB) AsyncFlush() error {
+func (d *DB) AsyncFlush() (chan struct{}, error) {
```
Comment (ajkr) [inline db.go]: I unified `AsyncFlush()` and `Flush()` because I thought `AsyncFlush()` would be needed for my test case. While it is still used there, I do not think it ended up being required, so we can revert these changes if they're unappealing.

File: db.go [hunk 2]
```diff
@@ -804,32 +802,48 @@ func (d *DB) walPreallocateSize() int {
 
 func (d *DB) makeRoomForWrite(b *Batch) error {
 	force := b == nil || b.flushable != nil
+	stalled := false
 	for {
```
Comment (ajkr) [inline db.go]: tbh I do not fully grok this loop. Is calling this function with `b == nil` supposed to always create a new WAL/memtable?

File: event_listener_test.go [hunk 3]
```diff
@@ -173,3 +174,75 @@ func TestEventListener(t *testing.T) {
...
+		MemTableStopWritesThreshold: 2,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// from now on delay writes until stall happens. Since WAL is disabled this results
+	// in delaying manifest writes, which blocks the flush thread. That causes memtables
+	// to accumulate, eventually resulting in a stall.
+	customFS.writesReleased = stallBegan
```
Comment (ajkr) [inline event_listener_test.go]: This assignment probably came too late as the manifest is already created. I think the test coincidentally works because the second call to `AsyncFlush()` will typically happen before the background flush completes. I will try adding some sleeps to verify there is indeed a race here.

To fix it, we can have the `writeDelayingFile` hold a pointer to the channel in `writeDelayingFS` (and create all files as `writeDelayingFile`s), so we can propagate the change instantaneously, rather than applying it only to files created after this line.

**Comment by ajkr** [comment]
> Sorry, though, I ended up slightly confused about both ideas. For the first one, would FlushBegin pause writes to the manifest the same way we're doing above?

Actually I think I got your first idea now. Will try it out.
